binary_search_tree/tree.py

```diff
@@ -9,51 +9,201 @@ def __init__(self, key, val = None):
+    # Space Complexity: O(1)/constant becuase nothing is created or removed
     def find(self, key):
-        pass
 
-    # Time Complexity: 
-    # Space Complexity: 
+        if self.root == None:
+            return None
+
+        current = self.root
+        while current != None:
+            if current.key == key:
+                return current.value
+            elif current.key < key:
+                current = current.right
+            else:
+                current = current.left
+
+        return None
+
+
+    def inorder_helper(self, current, nodes_array):
+        #Why do i need this helper to make this traversal work?
+        if current != None:
+            self.inorder_helper(current.left, nodes_array)
+            nodes_array.append({"key": current.key, "value": current.value})
+            self.inorder_helper(current.right, nodes_array)
+
+    # Time Complexity: O(n)/related to the number of nodes in the tree
+    # Space Complexity: O(1)/nothing is added
     def inorder(self):
-        pass
+    #left, root, right
+        nodes_array = []
+        self.inorder_helper(self.root, nodes_array)
+        return nodes_array 
 
-    # Time Complexity: 
-    # Space Complexity:     
+    def _preorder_helper(self, current, nodes_array):
+        if current != None:
+            nodes_array.append({"key": current.key, "value": current.value})
+            self._preorder_helper(current.left, nodes_array)
+            self._preorder_helper(current.right, nodes_array) 
+    
+    # Time Complexity: O(n)/related to the number of nodes in the tree
+    # Space Complexity: O(1)/nothing is added
     def preorder(self):
-        pass
+    #root, left, right
 
-    # Time Complexity: 
-    # Space Complexity:     
+    #create the nodes array here so that its only created here
+        nodes_array = []
+        self._preorder_helper(self.root, nodes_array)
+        return nodes_array
+
+    def _postorder_helper(self, current, nodes_array):
+        if current == None:
+            return None
+        
+        self._postorder_helper(current.left, nodes_array)
+        self._postorder_helper(current.right, nodes_array)
+        nodes_array.append({"key": current.key, "value": current.value})
+
+    # Time Complexity: O(n)/related to the number of nodes in the tree
+    # Space Complexity: O(1)/nothing is added   
     def postorder(self):
-        pass
+    #left, right, root
+        nodes_array = []
+        if self.root != None:
+            self._postorder_helper(self.root, nodes_array)
+        return nodes_array
+
+    def height_helper(self, current):
 
-    # Time Complexity: 
-    # Space Complexity:     
+        if current == None:
+            return 0
+        
+        return 1 + max(self.height_helper(current.left), self.height_helper(current.right))
+
+
+    # Time Complexity: O(n)/related to the number of nodes in the tree
+    # Space Complexity: O(1)/nothing is added  
     def height(self):
```

Comment:
👍  But space complexity is O(log n ) for a balanced tree due to the recursive call stack (and O(n) for an unbalanced tree).